src/hotspot/share/services/virtualMemoryTracker.cpp

```diff
@@ -523,29 +525,72 @@ bool VirtualMemoryTracker::remove_released_region(address addr, size_t size) {
     }
   }
 
-  VirtualMemorySummary::record_released_memory(size, reserved_rgn->flag());
-
-  assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
-  if (reserved_rgn->base() == addr ||
-      reserved_rgn->end() == addr + size) {
+  if (size <= reserved_rgn->size() && reserved_rgn->contain_region(addr, size)) {
+    // The requested chunk to be freed is smaller or same size as the real
+    // underlaying region and is fully contained within one (i.e. single region)
+    VirtualMemorySummary::record_released_memory(size, reserved_rgn->flag());
+    assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
+    if (reserved_rgn->base() == addr || reserved_rgn->end() == addr + size) {
       reserved_rgn->exclude_region(addr, size);
-    return true;
+      return true;
+    } else {
```

Comment:
OK! We split the region into two, that looks good. What happens to the committed regions that are part of both the removed reserved region and one of the surviving regions (or why not both?).

```
          *   &
          |   |
[---------------]
       [----]
~>
[-----]     [---]
```

What if we have a committed region starting by the `*` and ending at the `&`, what happens to it?

Reply:
Easier example:

```
- = reserved
C = committed
D = to be deleted
[---------------]
  [CCCCCCCCC]
     [DDD]
Are the results:
[----]   [-------]
  [CC]   [CC]
Where the left C is part of the left - and the right C is part of the right -
Because it looks like the results might be:
[----]   [-------]
  [CCCCCCCCC]
where the left - owns the committed region
```

Reply:
It looks like the tests only check the results of reserved regions.